near-sdk/src/json_types/hash.rs

```diff
@@ -60,6 +59,14 @@ impl TryFrom<&str> for Base58CryptoHash {
     type Error = Box<dyn std::error::Error>;
 
     fn try_from(value: &str) -> Result<Self, Self::Error> {
+        Self::from_str(value)
+    }
+}
+
+impl FromStr for Base58CryptoHash {
+    type Err = Box<dyn std::error::Error>;
```

Comment:
If this were a general-purpose probject, I'd write

```suggestion
    type Err = Box<dyn std::error::Error + Send + Sync>;
```

but I guess for us it doesn't really make sense

Reply:
I didn't want to open up to a breaking change, but I've created an alias to make this error type easier to switch in the future

Reply:
This particular impl is new, so, for this impl, we can change the error type, as long as it is convertable to original one via `?`. In particular, changing this (but only this) type to `ParseCryptoHashError` would work.

Reply:
which `ParseCryptoHashError` are you referring to? The box std::error with send/sync or are you referring to something specific/ generating a new error type?

Reply:
Sorry for being confusing here. I mean we can do the following, which should is backwards compatible, but, at the same time, is forwards compatible with a nice API of a hypothetical ideal sdk. 

```rust
impl TryFrom<&str> for Base58CryptoHash {
    type Error = Box<dyn std::error::Error>;
    type Error = Error;

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        let res = value.parse()?; // `?` converts ParseCryptoHashError to Box<dyn std::error::Error>
        Ok(res)
    }
}

#[non_exhaustive]
pub struct ParseCryptoHashError;
impl std::error::Error for ParseCryptoHashError { ... }

impl std::str::FromStr for Base58CryptoHash {
    // Ok to use new type here, as this is a new impl, so we don't care about backcompat. 
    type Err = ParseCryptoHashError;

    fn from_str(value: &str) -> Result<Self, ParseCryptoHashError> {
        ...
    }
}
```

Reply:
Yeah, the only reason I'm hesitant to add this is that it's unclear what the exact Error will be in the future. Using a concrete type like this enforces we keep that type (and trait bounds) in this ideal future API.

I'm happy to make the change and manually implement std::error if someone gives me the go-ahead with a thumbs up on this comment

Reply:
> is that it's unclear what the exact Error will be in the future

I think it's probably good to keep errors for well-isolated parts of code specific. IE, even we have something like `type NearError = ...` in the future, it would still make sense to use `ParseCryptoHashError` here. 

I don't think that'll affect the size of the compiled contract (the type is ZST, the implementation is trivial and inlinable, etc), though my confidence level here is not 100%.